### DISCUSSION TITLE
Wheels 3.0.0 SNAPSHOT is now available for testing

### DISCUSSION THREAD

**Discussion by bpamiri** [wheels-dev/wheels#1422] [Announcements]
Hello everyone,

We're finally ready for you to kick the tires on the prerelease version of Wheels 3.0. We now have snapshot releases posted to ForgeBox.io that can be installed via the CLI to get a preview look at the next version of Wheels.

Let's open a terminal window and start CommandBox by issuing a `box` command. Now we'll remove the old CLI tools, clear the CommandBox artifacts, and install the new CLI tools.

```
uninstall cfwheels-cli
artifacts remove cfwheels-cli
install wheels-cli
```

To insure that the CLI tools got installed properly you can use `help wheels` to see if the CLI tools got installed.

Now let's install a new Wheels application. You can use `wheels new` and go through the wizard or use the following command:
```
wheels g app name=wheels3 cfmlEngine=lucee template=wheels-base-template@be datasourceName=wheels3 reloadPassword= --setupH2
```
This command will create a folder named wheels3, download the latest Bleeding Edge version of the Wheels into that directory, set the application and datasource names to wheels3, and setup H2 as the database.

You can now issue a `server start` command

**Comment by dbelanger** [comment]
For the first time in a long time I finished early on Friday and was able to install and poke around the new wheels 3.0!

Love the new structure and I was able to use the CLI to install without issue.  My only suggestion is that instead of stopping the installation if the target directory is not empty, we warn people that the target files may be overwritten.  I only had a **.gitignore** file and my hidden **.git** folder so I was hoping to see the changes as they landed.  Still, it was easy to move the files from the default wheels3 directory to my intended directory after the fact as a workaround.  I only needed to update the **config/app.cfm** so that the h2 connection string matched my move.

Now, the first thing I did was to start the core tests.  The H2 driver was loaded as expected when I followed the CLI post install instructions of running "install" again after the server had loaded.  However I received a _"Datasource [wheelstestdb_h2] doesn't exist"_ for my model datasources test.

Turns out line 5 of `wheels/tests/model/read/datasource.cfc` has an alt datasource pointing to **wheelstestdb_h2**.  I changed it **wheels3** and all the tests come up green!!!

Next step this week - trying out our smallest app with wheels3 to see how it fairs.  Have we starting any docs for v3 yet?

**Reply by adamcameron** [reply]
I strongly agree that framework installers' habit of going "nuh-uh" if the target directory is not empty is... "unhelpful". Symfony does the same thing, and it's bloody annoying.

*My* approach would be to just get on with it, and only if there are file collisions do something about it. Options:
* exit out, to be on the safe side (stretch goal: backing out existing progress);
* have a JFDI _option_. Sling in the new files, bulldoze what's already there when there are conflicts;
* rename existent files that collide with CFWheels files, and emit that this has been done.

I will _always_ have git, docker, ide, etc files in a dir I install any third-party platform in.

NB: don't action any suggestion I make just cos I say it. I'll never be using 3.x, I'm just sharing my experiences with similar situations, and reflecting on how _I'd_ do it to be more helpful to the user.

**Reply by bpamiri** [reply]
Thanks for your input @adamcameron , it is much appreciated it. I appreciate people sharing their experiences with other environments and pointing out what they liked and what they didn't like about how the other environments handle certain situations. 

And I haven't given up on you yet my friend, who knows we may yet get a 3.0 app out of you. ;-)

**Comment by bpamiri** [comment]
@dbelanger great to hear you are playing with it. Negative on the docs as of yet. There are lots of changes that would have to be made and I wanted to make sure we get more feedback before I shifted gears. Personally I've tried taking an app we recently started using Example App to 3.0 and we ran into some CSRF errors. We've been investigating those for the last couple of week trying to get to the bottom of them. I also tried to do another very simple app in 3.0 and put it on a hosting plan (well it was a Plesk server running in my own datacenter) and I ran into issues with the public files being in a public folder. We may have to rethink that approach cause most hosting accounts will want the launch files to be in the root and not a sub directory. We may have to have an assets directory, and move the Application.cfc and index.cfm files to the root. I haven't decided the best approach for this yet. I was hoping to get some more feedback before making a commitment either way. This all points to the reason I haven't jumped into doc mode yet.

All in all glad to see someone kicking the tires.

**Comment by geirman** [comment]
I'm starting a new app. Just wondering if it would be advisable to continue using 2.5 or should I go ahead and start with 3.0?
I'm currently exploring the 'Example App' template to use as a basis because it has some pretty cool stuff in there already (auth, password reset, account verification, roles, logging, etc) that would give me a pretty nice jump start (if I can get it working, but that's for another post)

**Comment by bpamiri** [comment]
Chris,

Although the work on the current 3.0 branch has been in place for months, just recently we have begun some optimization work and there are some issues with Boxlang. So if you are going to go into production fairly soon, I would stick with the 2.5 branch till we get the rest of the work completed on the 3.0 branch.

--Peter Amiri
  ***@***.***

**Reply by geirman** [reply]
Thank you Peter, and good luck with your optimization efforts. I'm very happy to see CFWheels in your very capable, caring hands.

**Comment by bpamiri** [comment]
We now have some 3.0 specific documentation at:

https://guides.cfwheels.org/2.5.0/3.0.0-snapshot
